### ISSUE TITLE
Training losses do not converge with adding extra 3D heads

### ISSUE THREAD

**Issue by monstre0731** [ultralytics/ultralytics#9966]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

Thanks for your wonderful work. I am trying to extend YOLOv8 object detection task to a 3D version on a Monocular image. 
I added several branches in the module named 'Detect': with output channels as 3 for 3D dimension (weight, height and length) 1 for depth and  1 for angle(range in [-3.14, 3.14]).

I froze backbone, modules in heads part except those three branched I added.  I selected L1 and MSE as the loss functions. 
Here is the code I have added for the dimension, angle and depth branches:

```
        # added 
        c4 = max((16, ch[0] // 4, self.num_whl))  # whl
        c5 = max((16, ch[0] // 4, self.num_angle)) # angle
        c6 = max((16, ch[0] // 4, self.num_depth)) # depth
        self.cv2 = nn.ModuleList(nn.Sequential(Conv(x, c2, 3), Conv(c2, c2, 3), nn.Conv2d(c2, self.num_bbox * self.reg_max, 1)) for x in ch) # bbox
        self.cv3 = nn.ModuleList(nn.Sequential(Conv(x, c3, 3), Conv(c3, c3, 3), nn.Conv2d(c3, self.nc, 1)) for x in ch)
        self.cv4 = nn.ModuleList(nn.Sequential(Conv(x, c4, 3), Conv(c4, c4, 3), nn.Conv2d(c4, self.num_whl, 1)) for x in ch) # whl
        self.cv5 = nn.ModuleList(nn.Sequential(Conv(x, c5, 3), Conv(c5, c5, 3), nn.Conv2d(c5, self.num_angle, 1)) for x in ch) # angle
        self.cv6 = nn.ModuleList(nn.Sequential(Conv(x, c6, 3), Conv(c6, c6, 3), nn.Conv2d(c6, self.num_depth, 1)) for x in ch) # depth
        self.dfl = DFL(self.reg_max) if self.reg_max > 1 else nn.Identity()
        
        self.conv = nn.Conv2d(16, 1, 1, bias=False).requires_grad_(False)
```
   Here is the code I have added for calculating loss:
```
        # Dim loss whl
        target_whl /= stride_tensor
        loss[3] = self.mse(pred_whl, target_whl.to(dtype))
        # Angle loss 
        target_angle /= stride_tensor
        loss[4] = self.L1(pred_angle, target_angle.to(dtype))
        # depth loss 
        target_depth /= stride_tensor
        loss[5] = self.L1(pred_depth, target_depth.to(dtype)) 
```

The 2D part works fine, however, the training loss doesn't converge at all, any idea?  Thanks in advance.
<img width="1102" alt="Capture d’écran, le 2024-04-10 à 21 39 28" src="https://github.com/ultralytics/ultralytics/assets/70557428/388ed1da-0b57-4904-91ec-42edd1d89af1">

**Comment by Zephyr69** [comment]
I can see two likely problems. 
1. This is a significant shift in the task from 2d object detection. The previously learned features may not be very helpful for the new objectives, and the model only has two layers for each head to adapt. You shall consider training from scratch.
2. Make sure your (dim, angle, depth) labels are correctly processed in all augmentations. Otherwise, this will surely break the learning for these three heads.

**Comment by glenn-jocher** [comment]
Thank you for sharing these insights! Indeed, adapting a 2D detection network to perform 3D object detection can present unique challenges. 

👍 Training from scratch could be beneficial to allow the model to learn features that are more relevant for 3D detection tasks. This approach might lead to a better understanding and adaptation by the model for the dimensions, angles, and depths as opposed to relying solely on pre-learned 2D features.

🔍 Also, it's crucial to ensure that the labels for dimensions, angles, and depths are properly handled during all stages, including augmentations. Inconsistencies in label processing can significantly affect the model's learning capability for these new tasks.

A quick tip for label processing during augmentations might look something like this:
```python
# Example to adjust 'angle' label during an augmentation step
if 'rotate' in augmentations:
    target_angle += rotation_degree * np.pi / 180  # Adjusting the angle label
```
Applying careful attention to these aspects can enhance the model's ability to learn from the newly introduced data attributes effectively.

**Comment by monstre0731** [comment]
Hi,

Thanks for your replies. Sorry that I didn't mention that I had muted the augmentation in the training process by manually giving that  **self.augment  =  False**, I will try to only freeze the two layers for Bbox and Cls, and keep backbone's gradients updated. 

I'll update my tries here. Thanks.
```
    def build_transforms(self, hyp=None):
        """Builds and appends transforms to the list."""
        **self.augment = False**
        print(f'self.augment: {self.augment}')
        if self.augment:
            hyp.mosaic = hyp.mosaic if self.augment and not self.rect else 0.0
            hyp.mixup = hyp.mixup if self.augment and not self.rect else 0.0
            transforms = v8_transforms(self, self.imgsz, hyp)
        else:
            transforms = Compose([LetterBox(new_shape=(self.imgsz, self.imgsz), scaleup=False)])
        transforms.append(
            Format(
                bbox_format="xywh",
                normalize=True,
                return_mask=self.use_segments,
                return_keypoint=self.use_keypoints,
                return_obb=self.use_obb,
                return_whl=self.use_whl,  # detect_ddd
                batch_idx=True,
                mask_ratio=hyp.mask_ratio,
                mask_overlap=hyp.overlap_mask,
                bgr=hyp.bgr if self.augment else 0.0,  # only affect training.
            )
        )
        return transforms
```

**Comment by monstre0731** [comment]
Hi,

I have tried activating the backbone and muted the augmentation process, the losses cannot still converge. I also looked at the values of target_dimensions, target_angle and target_depth, they are fine.  

One point I cannot understand is that it failed with all three tasks.  More deeply, I notice that the duration of pred_depth is pretty smalll, from [0.6, 3.8], I didn't use any reg_max, but direct give output_channels as 3, 1, 1 for dimension, angle and depth. Will this be a problem causing it doesn't converge? Thanks again.

**Comment by glenn-jocher** [comment]
@monstre0731 hi there! 👋

Thanks for your update! If you've verified the target values and are still facing convergence issues even with the backbone gradients enabled and augmentations disabled, it's worth examining the scale of your output predictions, especially for depth. Given your predicted depth range [0.6, 3.8], if this range doesn't match the scale of your target depth values closely, it might indeed hinder convergence. 

One suggestion is to normalize your target depth values to a similar range as your predictions during training or vice versa. This alignment in scale could help improve learning dynamics.

For depth output without using `reg_max`, directly setting output channels works, but ensure the target values’ scale is consistent with model predictions. Sometimes applying a sigmoid function to restrict output range and then scaling it to the expected range of your target values can also help. Here's a quick snippet:
```python
# Assuming pred_depth is your model's raw output for depth
pred_depth = torch.sigmoid(pred_depth) * scale_factor  # scale_factor to match target depth range
```

Keeping a close tab on the scale and distribution of prediction and target values often unravels such convergence issues! Let us know how it goes. 😊

**Comment by monstre0731** [comment]
Hi, 

I use an activation function rely instead of sigmoid + scale_factor, without normalization depth to [-1, 1], the training depth losses decreased from 29 to 3. It seems like the loss decreasing but still not overfitting. 

Besides, I have a question about the stride_tensor in the loss.py, could you explain why the bboxes * stride_tensor, what is the motivation of stride_tensor and do I need to do the same for 3D assign for target_depth, target_angle etc? Thanks.
```
    _, target_bboxes, target_scores, fg_mask, _ = self.assigner(
        pred_scores.detach().sigmoid(),
        (pred_bboxes.detach() * stride_tensor).type(gt_bboxes.dtype),
        anchor_points * stride_tensor,
        gt_labels,
        gt_bboxes,
        mask_gt,
    )
```

**Comment by glenn-jocher** [comment]
@monstre0731 hey 😊,

Great to hear your depth losses have significantly decreased using ReLU! That's a positive shift.

Regarding the `stride_tensor`, it's there to scale back the prediction and anchor points to the original image dimensions. Since YOLOv8 operates on feature maps of different resolutions due to its backbone and neck design, `stride_tensor` helps in mapping the detections from the scaled-down feature maps back onto the original image size. This way, bbox predictions are in the same scale as the input image, making it easier to calculate the loss against ground truth annotations.

For your 3D attributes like depth and angle, if they're relative to image dimensions or require matching scales with the 2D bounding boxes, you might consider similar scaling adjustments. However, if these attributes are independent of image scale (for instance, if depth is an absolute measure not impacted by image resolution), additional scaling may not be necessary.

Let us know if this helps or if you have more questions!
